### PR TITLE
component: refactor set state handling

### PR DESCRIPTION
**Pull request by tlauda**
Changes component set state handling. If component
already has the desired state, then the state propagation
will stop at that component. This allows for connecting
multiple source pipelines to one sink pipeline and
properly handling params and prepare calls.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

**Comment by lgirdwood** [comment]
@tlauda looks like some UTs have failed.

File: src/audio/component.c [hunk 1]
```diff
@@ -119,8 +119,15 @@ void comp_unregister(struct comp_driver *drv)
...
 int comp_set_state(struct comp_dev *dev, int cmd)
 {
+	int requested_state = comp_get_requested_state(cmd);
 	int ret = 0;
 
+	if (dev->state == requested_state) {
+		trace_comp("comp_set_state(), state already set to %u",
+			   dev->state);
+		return 1;
+	}
```
Comment (dbaluta) [inline src/audio/component.c]: I wonder if it wouldn't be better to just return 0 here. There is no distinction in the code between "state was changed" and "state is already as it should be".

With this, we only have two possible values for comp_set_state returned value: 0 for success, negative value for error.

Comment (tlauda) [reply src/audio/component.c]: It returns 1 on purpose to stop walking pipeline graph any further. It is assumed, that current pipeline has been already set to the desired state. Desired in scenario with multiple connected pipelines. Probably such handling should be revisited and refactored in the future.

Comment (dbaluta) [reply src/audio/component.c]: @tlauda makes sense. Thanks for the explanation.

**Comment by tlauda** [comment]
@jajanusz Yes, as stated in commit message.

**Comment by jajanusz** [comment]
> @jajanusz Yes, as stated in commit message.

Sorry, didn't read commit description :man_facepalming:

**Pull request merged**